rpctest/rpcharness.go

```diff
@@ -0,0 +1,416 @@
+// license that can be found in the LICENSE file.
+
+package rpctest
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrwallet/chain"
+
+	rpc "github.com/decred/dcrrpcclient"
+	"github.com/decred/dcrutil"
+)
+
+var (
+	// current number of active test nodes.
+	numTestInstances = 0
+
+	// defaultP2pPort is the initial p2p port which will be used by the
+	// first created rpc harnesses to listen on for incoming p2p connections.
+	// Subsequent allocated ports for future rpc harness instances will be
+	// monotonically increasing odd numbers calculated as such:
+	// defaultP2pPort + (2 * harness.nodeNum).
+	defaultP2pPort = 18555
+
+	// defaultRPCPort is the initial rpc port which will be used by the first created
+	// rpc harnesses to listen on for incoming rpc connections. Subsequent
+	// allocated ports for future rpc harness instances will be monotonically
+	// increasing even numbers calculated as such:
+	// defaultP2pPort + (2 * harness.nodeNum).
+	defaultRPCPort = 19556
+
+	defaultWalletRPCPort = 19557
+
+	// testInstances is a private package-level slice used to keep track of
+	// allvactive test harnesses. This global can be used to perform various
+	// "joins", shutdown several active harnesses after a test, etc.
+	testInstances map[string]*Harness
+
+	// Used to protest concurrent access to above declared variables.
+	harnessStateMtx sync.RWMutex
+)
+
+// Harness fully encapsulates an active dcrd process, along with an embdedded
+// dcrwallet in order to provide a unified platform for creating rpc driven
+// integration tests involving dcrd. The active dcrd node will typically be
+// run in simnet mode in order to allow for easy generation of test blockchains.
+// Additionally, a special method is provided which allows on to easily generate
+// coinbase spends. The active dcrd process if fully managed by Harness, which
+// handles the necessary initialization, and teardown of the process along with
+// any temporary directories created as a result. Multiple Harness instances may
+// be run concurrently, in order to allow for testing complex scenarios involving
+// multuple nodes.
+type Harness struct {
+	ActiveNet *chaincfg.Params
+
+	Node      *rpc.Client
+	WalletRPC *rpc.Client
+	node      *node
+	handlers  *rpc.NotificationHandlers
+
+	wallet *walletTest
+
+	chainClient *chain.RPCClient
+
+	testNodeDir    string
+	testWalletDir  string
+	maxConnRetries int
+	nodeNum        int
+}
+
+// New creates and initializes new instance of the rpc test harness.
+// Optionally, websocket handlers and a specified configuration may be passed.
+// In the case that a nil config is passed, a default configuration will be used.
+//
+// NOTE: This function is safe for concurrent access.
+func New(activeNet *chaincfg.Params, handlers *rpc.NotificationHandlers,
```

Comment:
Constructor should be of the form "NewHarness" to be consistent with the rest of the codebase